src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,280 @@ impl<F: Field> ConstraintSystem<F> {
+
+                // Keep track of how large the substitution is.
+                let mut usage = 1;
+
+                // Keep track of how large of a gate any of the members of this
+                // combination can support.
+                let mut room = max_degree - degrees[i];
+
+                // Try to find other selectors that can join this one.
+                for j in (i + 1)..self.num_selectors {
+                    // Is there room for the new selector?
+                    if room == 0 {
+                        // Short circuit here, we've hit our limit.
+                        break;
+                    }
+
+                    // Skip complex selectors.
+                    if complex_selectors[j] {
+                        continue;
+                    }
+
+                    // Skip selectors that have been added to previous combinations
+                    if seen[j] {
+                        continue;
+                    }
+
+                    // Are the two selectors excluded from co-existing?
+                    // j > i and so (j, i) is always on the lower diagonal
+                    if exclusion_matrix[j][i] {
+                        continue;
+                    }
+
+                    // Can the new selector join the combination?
+                    if degrees[j] + usage <= max_degree {
+                        // Yes, it can. Let's do it.
+                        combination.push(j);
+                        // The remaining room available is the lesser of the
+                        // current remaining room available for previous
+                        // selector(s) after adjusting for adding this new
+                        // selector, and the room available in the new selector
+                        // after adjusting for adding the previous selectors.
+                        room = std::cmp::min(room - 1, max_degree - (degrees[j] + usage));
+                        usage += 1;
+                        selector_map[j] = combinations.len();
+                        seen[j] = true;
+                    }
+                }
+
+                combinations.push(combination);
+            }
+        }
+
+        let mut selector_replacements = vec![None; self.num_selectors];
+        let mut polys = vec![vec![F::zero(); n]; combinations.len()];
+        let mut new_columns = vec![];
+        for (combination, poly) in combinations.iter().zip(polys.iter_mut()) {
+            // Create a new fixed column for the logical selector
+            let column = self.fixed_column();
+            new_columns.push(Some(column));
+            let query = Expression::Fixed {
+                query_index: self.query_fixed_index(column, Rotation::cur()),
+                column_index: column.index,
+                rotation: Rotation::cur(),
+            };
+
+            // Every virtual selector gets a different integer (starting at 1)
+            // that it assigns when it wants to be active.
```

Comment:
```suggestion
            // Every virtual selector in this combination gets a different integer
            // (starting at 1) that is assigned when it is enabled.
```
(You *could* assign disjoint integers across all combinations. I guess that might have minor advantages for debugging, but it would be very slightly less efficient, due to larger scalars in the multiexp.)